workflow/controller/controller.go

```diff
@@ -1525,3 +1538,14 @@ func (wfc *WorkflowController) IsLeader() bool {
 	// the wfc.wfInformer is nil if it is not the leader
 	return !(wfc.wfInformer == nil)
 }
+
+func (wfc *WorkflowController) WatchErrorHandler(r *cache.Reflector, err error) {
+	cache.DefaultWatchErrorHandler(r, err)
+	if err != io.EOF {
+		// The reflector will set the Limit to `0` when `ResourceVersion != "" && ResourceVersion != "0"`, which will fail
+		// to limit the number of workflow returns. Timeouts and other errors may occur when there are a lots of workflows.
+		// see https://github.com/kubernetes/client-go/blob/ee1a5aaf793a9ace9c433f5fb26a19058ed5f37c/tools/cache/reflector.go#L286
+		r.WatchListPageSize = common.DefaultPageSize
+		wfc.WatchListPageSize = r.WatchListPageSize
+	}
+}
```

Comment:
I added an error handler to set the `WatchListPageSize` of the `reflector` to ensure that it can meet the Limit restriction when an error occurs in `ListAndWatch`. 
Currently, I have only found this way to set `WatchListPageSize`. If anyone has a better solution, I am happy to discuss it with you.